# PyCity Schools Analysis

In [101]:
# Dependencies and Setup
import pandas as pd
from pathlib import Path

# File to Load (Remember to Change These)
school_data_to_load = Path("../Resources/schools_complete.csv")
student_data_to_load = Path("../Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
schools_df = pd.read_csv(school_data_to_load)
student_df = pd.read_csv(student_data_to_load)

# Combine the data into a single dataset.  
data_complete = pd.merge(student_df, schools_df, how="outer", on=["school_name", "school_name"])

Huang, Figueroa, Shelton, Hernandez, Griffin, Wilson, Cabrera, Bailey, Holden, Pena, Wright, Rodriguez, Johnson, Ford, Thomas

There are 15 schools - store in a variable

In [102]:
school_count = len(schools_df["school_name"])
student_count = len(student_df["Student ID"].unique())
print(student_count)

budgets_all = schools_df["budget"].unique()
print(budgets_all)

budgets_total = budgets_all.sum()
print(budgets_total)

39170
[1910635 1884411 1056600 3022020  917500 1319574 1081356 3124928  248087
  585858 1049400 2547363 3094650 1763916 1043130]
24649428


In [103]:
avg_reading_score = data_complete["reading_score"].mean()
print(avg_reading_score)

avg_math_score = data_complete["math_score"].mean()
print(avg_math_score)

81.87784018381414
78.98537145774827


data_complete["Student ID"].value_counts()
data_complete.describe()

this shows 39,170 students

length of the Student ID column is 39170   same for all columns

get enrollment count at each school from the size column

overall average math score stored in avg_math_score
overall average reading score stored in ave_reading_score

Creating a new column for pass or fail math score with 70 and higher being pass and lower being fail
and one for reading also

In [104]:

data_complete["math_pass"] = data_complete["math_score"].apply(lambda x:"pass" if x>=70 else "fail")

data_complete["reading_pass"] = data_complete["reading_score"].apply(lambda x:"pass" if x>=70 else "fail")

math_num_pass = data_complete["math_pass"].value_counts()
reading_num_pass = data_complete["reading_pass"].value_counts()

math_percent_pass = (math_num_pass / student_count)*100
reading_percent_pass = (reading_num_pass / student_count)*100

def get_status(row):
    if row["math_pass"] == "pass" and row["reading_pass"] == "pass":
        return "pass"
    else:
        return "fail"

data_complete["overall pass"] = data_complete.apply(lambda row: get_status(row), axis=1)

overall_num_pass = data_complete["overall pass"].value_counts()

overall_percent_pass = (overall_num_pass / student_count)*100

print("District Summary")
print("There are ", school_count, "schools in the district.")
print("There are ", student_count, "total students.")
print("The total district budget is $", budgets_total, ".")
print("The average math score for the district is ", avg_math_score, ".")
print("The average reading score for the district is ", avg_reading_score, ".")
print("Percentage of students passing math is ", math_percent_pass, "%.")
print("Percentage of students passing reading is ", reading_percent_pass, "%.")
print("Percentage of students passing both math and reading is ", overall_percent_pass, "%.")



District Summary
There are  15 schools in the district.
There are  39170 total students.
The total district budget is $ 24649428 .
The average math score for the district is  78.98537145774827 .
The average reading score for the district is  81.87784018381414 .
Percentage of students passing math is  math_pass
pass    74.980853
fail    25.019147
Name: count, dtype: float64 %.
Percentage of students passing reading is  reading_pass
pass    85.805463
fail    14.194537
Name: count, dtype: float64 %.
Percentage of students passing both math and reading is  overall pass
pass    65.172326
fail    34.827674
Name: count, dtype: float64 %.


Make a dataframe for each individual school

In [105]:
district_summary = pd.DataFrame([{"Total Schools":school_count,"Total Students":student_count,
                                  "Total Budget":budgets_total,"Average Math Score": avg_math_score,
                                  "Average Reading Score":avg_reading_score,"% Passing Math":math_percent_pass,
                                  "% Passing Reading":reading_percent_pass,"% Overall Passing":overall_percent_pass}])

district_summary["Total Students"] = district_summary["Total Students"].map("{:,}".format)
district_summary["Total Budget"] = district_summary["Total Budget"].map("${:,.2f}".format)

In [106]:
students_passing_math_and_reading = data_complete[
    (data_complete["reading_score"] >= 70) & (data_complete["math_score"] >= 70)
]
school_students_passing_math_and_reading = students_passing_math_and_reading.groupby(["school_name"]).size()

school_students_passing_math_and_reading_df = pd.DataFrame(school_students_passing_math_and_reading[[""]]

school_name
Bailey High School       2719
Cabrera High School      1697
Figueroa High School     1569
Ford High School         1487
Griffin High School      1330
Hernandez High School    2481
Holden High School        381
Huang High School        1561
Johnson High School      2549
Pena High School          871
Rodriguez High School    2119
Shelton High School      1583
Thomas High School       1487
Wilson High School       2068
Wright High School       1626
dtype: int64

school_name_df = data_complete.set_index("school_name")

Huang_df = school_name_df.loc["Huang High School", :]
Hu_math_count=Huang_df['math_pass'].str.contains('pass', case=False).sum()
Hu_read_count=Huang_df['reading_pass'].str.contains('pass', case=False).sum()
Hu_all_count=Huang_df['overall pass'].str.contains('pass', case=False).sum()

Bailey_df = school_name_df.loc["Bailey High School", :]
Ba_math_count=Bailey_df['math_pass'].str.contains('pass', case=False).sum()
Ba_read_count=Bailey_df['reading_pass'].str.contains('pass', case=False).sum()
Ba_all_count=Bailey_df['overall pass'].str.contains('pass', case=False).sum()

Cabrera_df = school_name_df.loc["Cabrera High School", :]
Cabrera_df_math_count=Cabrera_df['math_pass'].str.contains('pass', case=False).sum()
Ca_read_count=Cabrera_df['reading_pass'].str.contains('pass', case=False).sum()
Ca_all_count=Cabrera_df['overall pass'].str.contains('pass', case=False).sum()

Figueroa_df = school_name_df.loc["Figueroa High School", :]
Fi_math_count=Figueroa_df['math_pass'].str.contains('pass', case=False).sum()
Fi_read_count=Figueroa_df['reading_pass'].str.contains('pass', case=False).sum()
Fi_all_count=Figueroa_df['overall pass'].str.contains('pass', case=False).sum()

Ford_df = school_name_df.loc["Bailey High School", :]
Fo_math_count=Ford_df['math_pass'].str.contains('pass', case=False).sum()
Fo_read_count=Ford_df['reading_pass'].str.contains('pass', case=False).sum()
Fo_all_count=Ford_df['overall pass'].str.contains('pass', case=False).sum()

Griffin_df = school_name_df.loc["Griffin High School", :]
Gr_math_count=Griffin_df['math_pass'].str.contains('pass', case=False).sum()
Gr_read_count=Griffin_df['reading_pass'].str.contains('pass', case=False).sum()
Gr_all_count=Griffin_df['overall pass'].str.contains('pass', case=False).sum()

Hernandez_df = school_name_df.loc["Hernandez High School", :]
He_math_count=Hernandez_df['math_pass'].str.contains('pass', case=False).sum()
He_read_count=Hernandez_df['reading_pass'].str.contains('pass', case=False).sum()
He_all_count=Hernandez_df['overall pass'].str.contains('pass', case=False).sum()

Holden_df = school_name_df.loc["Holden High School", :]
Ho_math_count=Holden_df['math_pass'].str.contains('pass', case=False).sum()
Ho_read_count=Holden_df['reading_pass'].str.contains('pass', case=False).sum()
Ho_all_count=Holden_df['overall pass'].str.contains('pass', case=False).sum()

print(Ho_all_count)


Johnson_df = school_name_df.loc["Johnson High School", :]
Jo_math_count=Johnson_df['math_pass'].str.contains('pass', case=False).sum()
Jo_read_count=Johnson_df['reading_pass'].str.contains('pass', case=False).sum()
Jo_all_count=Johnson_df['overall pass'].str.contains('pass', case=False).sum()

Pena_df = school_name_df.loc["Pena High School", :]
Pe_math_count=Pena_df['math_pass'].str.contains('pass', case=False).sum()
Pe_read_count=Pena_df['reading_pass'].str.contains('pass', case=False).sum()
Pe_all_count=Pena_df['overall pass'].str.contains('pass', case=False).sum()

Rodriguez_df = school_name_df.loc["Rodriguez High School", :]
Ro_math_count=Rodriguez_df['math_pass'].str.contains('pass', case=False).sum()
Ro_read_count=Rodriguez_df['reading_pass'].str.contains('pass', case=False).sum()
Ro_all_count=Rodriguez_df['overall pass'].str.contains('pass', case=False).sum()

Shelton_df = school_name_df.loc["Shelton High School", :]
Sh_math_count=Shelton_df['math_pass'].str.contains('pass', case=False).sum()
Sh_read_count=Shelton_df['reading_pass'].str.contains('pass', case=False).sum()
Sh_all_count=Shelton_df['overall pass'].str.contains('pass', case=False).sum()

Thomas_df = school_name_df.loc["Thomas High School", :]
Th_math_count=Thomas_df['math_pass'].str.contains('pass', case=False).sum()
Th_read_count=Thomas_df['reading_pass'].str.contains('pass', case=False).sum()
Th_all_count=Thomas_df['overall pass'].str.contains('pass', case=False).sum()

Wilson_df = school_name_df.loc["Wilson High School", :]
Wi_math_count=Wilson_df['math_pass'].str.contains('pass', case=False).sum()
Wi_read_count=Wilson_df['reading_pass'].str.contains('pass', case=False).sum()
Wi_all_count=Wilson_df['overall pass'].str.contains('pass', case=False).sum()

Wright_df = school_name_df.loc["Wright High School", :]
Wr_math_count=Wright_df['math_pass'].str.contains('pass', case=False).sum()
Wr_read_count=Wright_df['reading_pass'].str.contains('pass', case=False).sum()
Wr_all_count=Wright_df['overall pass'].str.contains('pass', case=False).sum()

print(Wr_all_count)

grouped_school_totals = data_complete.groupby(["school_name","math_pass"])
math_pass_school_df = pd.DataFrame(grouped_school_totals[["math_pass"]].count())

grouped_school_totals2 = data_complete.groupby(["school_name","reading_pass"])
reading_pass_school_df = pd.DataFrame(grouped_school_totals2[["reading_pass"]].count())

grouped_school_totals3 = data_complete.groupby(["school_name","overall pass"])
overall_pass_school_df = pd.DataFrame(grouped_school_totals2[["overall pass"]].count())

math_pass_school1_df = math_pass_school_df.rename(columns = {"math_pass":"math_pass_or_fail"})
read_pass_school1_df = reading_pass_school_df.rename(columns = {"reading_pass":"reading_pass_or_fail"})
math_pass_school1_df


read_pass_lib=[read_pass_school1_df.iloc[1,0],read_pass_school1_df.iloc[3,0],read_pass_school1_df.iloc[5,0],read_pass_school1_df.iloc[7,0],
           read_pass_school1_df.iloc[9,0],read_pass_school1_df.iloc[11,0],read_pass_school1_df.iloc[13,0],read_pass_school1_df.iloc[15,0],
           read_pass_school1_df.iloc[17,0],read_pass_school1_df.iloc[19,0],read_pass_school1_df.iloc[21,0],read_pass_school1_df.iloc[23,0],
           read_pass_school1_df.iloc[25,0],read_pass_school1_df.iloc[27,0],read_pass_school1_df.iloc[29,0]]
math_pass_lib=[math_pass_school1_df.iloc[1,0],math_pass_school1_df.iloc[3,0],math_pass_school1_df.iloc[5,0],math_pass_school1_df.iloc[7,0],
           math_pass_school1_df.iloc[9,0],math_pass_school1_df.iloc[11,0],math_pass_school1_df.iloc[13,0],math_pass_school1_df.iloc[15,0],
           math_pass_school1_df.iloc[17,0],math_pass_school1_df.iloc[19,0],math_pass_school1_df.iloc[21,0],math_pass_school1_df.iloc[23,0],
           math_pass_school1_df.iloc[25,0],math_pass_school1_df.iloc[27,0],math_pass_school1_df.iloc[29,0]]
overall_pass_lib=[overall_pass_school_df.iloc[1,0],overall_pass_school_df.iloc[3,0],overall_pass_school_df.iloc[5,0],overall_pass_school_df.iloc[7,0],
           overall_pass_school_df.iloc[9,0],overall_pass_school_df.iloc[11,0],overall_pass_school_df.iloc[13,0],overall_pass_school_df.iloc[15,0],
           overall_pass_school_df.iloc[17,0],overall_pass_school_df.iloc[19,0],overall_pass_school_df.iloc[21,0],overall_pass_school_df.iloc[23,0],
           overall_pass_school_df.iloc[25,0],overall_pass_school_df.iloc[27,0],overall_pass_school_df.iloc[29,0]]
name_lib = ['Bailey High School', 'Cabrera High School','Figueroa High School', 'Ford High School','Griffin High School','Hernandez High School', 'Holden High School', 'Huang High School', 'Johnson High School', 'Pena High School', 'Rodriguez High School', 'Shelton High School', 'Thomas High School', 'Wilson High School', 'Wright High School']
school_pass_df = pd.DataFrame({"school_name":name_lib,"Number Passing Math":math_pass_lib, "Number Passing Reading":read_pass_lib, "Overall Number Passing":overall_pass_lib})
school_pass_df

pass_merge_df = pd.merge(schools_df, math_pass_school1_df, on= "school_name", how = "inner")
pass_merge2_df = pd.merge(pass_merge_df,read_pass_school1_df, on="school_name",how = "inner")
pass_merge2_df

data_complete["overall pass"] = data_complete.apply(lambda row: get_status(row), axis=1)

overall_num_pass = data_complete["overall pass"].value_counts()

overall_percent_pass = (overall_num_pass / student_count)*100

overall_percent_pass.head()


In [111]:

data_complete["overall pass"] = data_complete.apply(lambda row: get_status(row), axis=1)

overall_num_pass = data_complete["overall pass"].value_counts()

overall_percent_pass = (overall_num_pass / student_count)*100

# drop the duplicate rows with the smaller number - the fails
pass_drop_df=pass_merge2_df.drop_duplicates(subset="school_name", keep="last")

renamed_pass_df = pass_drop_df.rename(columns={"math_pass_or_fail":"Num_pass_math","reading_pass_or_fail":"Num_pass_read"})

renamed_pass_df["% passing math"]=renamed_pass_df["Num_pass_math"]/renamed_pass_df["size"]*100
renamed_pass_df["% passing reading"]=renamed_pass_df["Num_pass_read"]/renamed_pass_df["size"]*100



#add column for per student budget
renamed_pass_df["Per student budget"]=renamed_pass_df["budget"]/renamed_pass_df["size"]


new_pass_df = pd.merge(renamed_pass_df,school_students_passing_math_and_reading, on="school_name")
new_pass_df


ValueError: Cannot merge a Series without a name

In [43]:
avg_school_scorem = data_complete.groupby(["school_name"])
avg_math_scores_df = pd.DataFrame(avg_school_scorem[["math_score"]].mean())

avg_read_scores_df =pd.DataFrame(avg_school_scorem[["reading_score"]].mean())

merge_scores_df = pd.merge(avg_math_scores_df,avg_read_scores_df,on = "school_name")

average_scores_df = merge_scores_df.rename(columns={"math_score":"Average math score","reading_score":"Average reading score"})
average_scores_df

Average math score  Average reading score
school_name                                                     
Bailey High School              77.048432              81.033963
Cabrera High School             83.061895              83.975780
Figueroa High School            76.711767              81.158020
Ford High School                77.102592              80.746258
Griffin High School             83.351499              83.816757
Hernandez High School           77.289752              80.934412
Holden High School              83.803279              83.814988
Huang High School               76.629414              81.182722
Johnson High School             77.072464              80.966394
Pena High School                83.839917              84.044699
Rodriguez High School           76.842711              80.744686
Shelton High School             83.359455              83.725724
Thomas High School              83.418349              83.848930
Wilson High School              83.274201              83.989488
Wright High School              83.682222              83.955000

In [44]:
final_merge_df = pd.merge(renamed_pass_df, average_scores_df, on= "school_name", how = "outer")
final_merge_df

School ID            school_name      type  size   budget  Num_pass_math  \
0           0      Huang High School  District  2917  1910635           1916   
1           1   Figueroa High School  District  2949  1884411           1946   
2           2    Shelton High School   Charter  1761  1056600           1653   
3           3  Hernandez High School  District  4635  3022020           3094   
4           4    Griffin High School   Charter  1468   917500           1371   
5           5     Wilson High School   Charter  2283  1319574           2143   
6           6    Cabrera High School   Charter  1858  1081356           1749   
7           7     Bailey High School  District  4976  3124928           3318   
8           8     Holden High School   Charter   427   248087            395   
9           9       Pena High School   Charter   962   585858            910   
10         10     Wright High School   Charter  1800  1049400           1680   
11         11  Rodriguez High School  District  3999  2547363           2654   
12         12    Johnson High School  District  4761  3094650           3145   
13         13       Ford High School  District  2739  1763916           1871   
14         14     Thomas High School   Charter  1635  1043130           1525   

    Num_pass_read  % passing math  % passing reading  Per student budget  \
0            2372       65.683922          81.316421               655.0   
1            2381       65.988471          80.739234               639.0   
2            1688       93.867121          95.854628               600.0   
3            3748       66.752967          80.862999               652.0   
4            1426       93.392371          97.138965               625.0   
5            2204       93.867718          96.539641               578.0   
6            1803       94.133477          97.039828               582.0   
7            4077       66.680064          81.933280               628.0   
8             411       92.505855          96.252927               581.0   
9             923       94.594595          95.945946               609.0   
10           1739       93.333333          96.611111               583.0   
11           3208       66.366592          80.220055               637.0   
12           3867       66.057551          81.222432               650.0   
13           2172       68.309602          79.299014               644.0   
14           1591       93.272171          97.308869               638.0   

    Average math score  Average reading score  
0            76.629414              81.182722  
1            76.711767              81.158020  
2            83.359455              83.725724  
3            77.289752              80.934412  
4            83.351499              83.816757  
5            83.274201              83.989488  
6            83.061895              83.975780  
7            77.048432              81.033963  
8            83.803279              83.814988  
9            83.839917              84.044699  
10           83.682222              83.955000  
11           76.842711              80.744686  
12           77.072464              80.966394  
13           77.102592              80.746258  
14           83.418349              83.848930

Now cleaning the dataframe, renaming columnns, organizing, deleting unneeded columns
first - indexing by school name

In [45]:
full_school_df = final_merge_df.set_index("school_name")
del full_school_df["School ID"]
renamed_df = full_school_df.rename(columns = {"type":"School Type", "size":"Total Students", "budget":"Total school budget", "math_score":"Average math score",
                                       "reading_score":"Average reading score"})
organized_df=renamed_df[["School Type","Total Students", "Total school budget", "Per student budget", "Average math score", "Average reading score", "% passing math", "% passing reading"]]
organized_df

School Type  Total Students  Total school budget  \
school_name                                                              
Huang High School        District            2917              1910635   
Figueroa High School     District            2949              1884411   
Shelton High School       Charter            1761              1056600   
Hernandez High School    District            4635              3022020   
Griffin High School       Charter            1468               917500   
Wilson High School        Charter            2283              1319574   
Cabrera High School       Charter            1858              1081356   
Bailey High School       District            4976              3124928   
Holden High School        Charter             427               248087   
Pena High School          Charter             962               585858   
Wright High School        Charter            1800              1049400   
Rodriguez High School    District            3999              2547363   
Johnson High School      District            4761              3094650   
Ford High School         District            2739              1763916   
Thomas High School        Charter            1635              1043130   

                       Per student budget  Average math score  \
school_name                                                     
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Shelton High School                 600.0           83.359455   
Hernandez High School               652.0           77.289752   
Griffin High School                 625.0           83.351499   
Wilson High School                  578.0           83.274201   
Cabrera High School                 582.0           83.061895   
Bailey High School                  628.0           77.048432   
Holden High School                  581.0           83.803279   
Pena High School                    609.0           83.839917   
Wright High School                  583.0           83.682222   
Rodriguez High School               637.0           76.842711   
Johnson High School                 650.0           77.072464   
Ford High School                    644.0           77.102592   
Thomas High School                  638.0           83.418349   

                       Average reading score  % passing math  \
school_name                                                    
Huang High School                  81.182722       65.683922   
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   
Wilson High School                 83.989488       93.867718   
Cabrera High School                83.975780       94.133477   
Bailey High School                 81.033963       66.680064   
Holden High School                 83.814988       92.505855   
Pena High School                   84.044699       94.594595   
Wright High School                 83.955000       93.333333   
Rodriguez High School              80.744686       66.366592   
Johnson High School                80.966394       66.057551   
Ford High School                   80.746258       68.309602   
Thomas High School                 83.848930       93.272171   

                       % passing reading  
school_name                               
Huang High School              81.316421  
Figueroa High School           80.739234  
Shelton High School            95.854628  
Hernandez High School          80.862999  
Griffin High School            97.138965  
Wilson High School             96.539641  
Cabrera High School            97.039828  
Bailey High School             81.933280  
Holden High School             96.252927  
Pena High School               95.945946  
Wright High School             96.611111  
Rodriguez

formatting values in this data frame
code is commented out because numerical values are needed for sorting and grouping later.  Copy this code to the end and adjust to format final results data frames.

def format_int():
    return "{:,}".format()

def format_per():
    return "{:.1f}%".format()

def format_budg():
    return "${:,}".format()

def format_psbudget():
    return "${.0f}".format()

organized_df.loc[:,"Total Students"] = organized_df.loc[:,"Total Students"].apply(format_int)


In [46]:
organized_df["% overall passing"]=organized_df["% passing math"]
organized_df.head()

C:\Users\mrkol\AppData\Local\Temp\ipykernel_22380\1385722768.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  organized_df["% overall passing"]=organized_df["% passing math"]


School Type  Total Students  Total school budget  \
school_name                                                              
Huang High School        District            2917              1910635   
Figueroa High School     District            2949              1884411   
Shelton High School       Charter            1761              1056600   
Hernandez High School    District            4635              3022020   
Griffin High School       Charter            1468               917500   

                       Per student budget  Average math score  \
school_name                                                     
Huang High School                   655.0           76.629414   
Figueroa High School                639.0           76.711767   
Shelton High School                 600.0           83.359455   
Hernandez High School               652.0           77.289752   
Griffin High School                 625.0           83.351499   

                       Average reading score  % passing math  \
school_name                                                    
Huang High School                  81.182722       65.683922   
Figueroa High School               81.158020       65.988471   
Shelton High School                83.725724       93.867121   
Hernandez High School              80.934412       66.752967   
Griffin High School                83.816757       93.392371   

                       % passing reading  % overall passing  
school_name                                                  
Huang High School              81.316421          65.683922  
Figueroa High School           80.739234          65.988471  
Shelton High School            95.854628          93.867121  
Hernandez High School          80.862999          66.752967  
Griffin High School            97.138965          93.392371

Now sort the school information.  % overall passing must not be mapped yet, so that it can be used to sort numerically

In [596]:
top_schools = organized_df.sort_values(by=["% overall passing","Total Students"], ascending=False).head()
top_schools["Total Students"] = top_schools["Total Students"].map("{:,}".format)
top_schools["% passing math"] = top_schools["% passing math"].map("{:.1f}%".format)
top_schools["% overall passing"] = top_schools["% overall passing"].map("{:.1f}%".format)
top_schools["Total school budget"] = top_schools["Total school budget"].map("${:,}".format)
top_schools["% passing reading"] = top_schools["% passing reading"].map("{:.1f}%".format)
top_schools["Average reading score"] = top_schools["Average reading score"].map("{:.1f}".format)
top_schools["Average math score"] = top_schools["Average math score"].map("{:.1f}".format)
top_schools

School Type Total Students Total school budget  \
school_name                                                          
Wilson High School      Charter          2,283          $1,319,574   
Cabrera High School     Charter          1,858          $1,081,356   
Wright High School      Charter          1,800          $1,049,400   
Shelton High School     Charter          1,761          $1,056,600   
Thomas High School      Charter          1,635          $1,043,130   

                     Per student budget Average math score  \
school_name                                                  
Wilson High School                578.0               83.3   
Cabrera High School               582.0               83.1   
Wright High School                583.0               83.7   
Shelton High School               600.0               83.4   
Thomas High School                638.0               83.4   

                    Average reading score % passing math % passing reading  \
school_name                                                                  
Wilson High School                   84.0         100.0%            100.0%   
Cabrera High School                  84.0         100.0%            100.0%   
Wright High School                   84.0         100.0%            100.0%   
Shelton High School                  83.7         100.0%            100.0%   
Thomas High School                   83.8         100.0%            100.0%   

                    % overall passing  
school_name                            
Wilson High School             100.0%  
Cabrera High School            100.0%  
Wright High School             100.0%  
Shelton High School            100.0%  
Thomas High School             100.0%

the top schools are organized by % overall passing, and then by Total Students in descending order, as larger schools have lower passing rates in general.  It is most impressive to achieve 100% passing for a large school

In [597]:
bottom_schools = organized_df.sort_values(by=["% overall passing"], ascending=True).head()
bottom_schools["Total Students"] = bottom_schools["Total Students"].map("{:,}".format)
bottom_schools["% passing math"] = bottom_schools["% passing math"].map("{:.1f}%".format)
bottom_schools["% overall passing"] = bottom_schools["% overall passing"].map("{:.1f}%".format)
bottom_schools["Total school budget"] = bottom_schools["Total school budget"].map("${:,}".format)
bottom_schools["% passing reading"] = bottom_schools["% passing reading"].map("{:.1f}%".format)
bottom_schools["Average reading score"] = bottom_schools["Average reading score"].map("{:.1f}".format)
bottom_schools["Average math score"] = bottom_schools["Average math score"].map("{:.1f}".format)
bottom_schools

School Type Total Students Total school budget  \
school_name                                                            
Figueroa High School     District          2,949          $1,884,411   
Rodriguez High School    District          3,999          $2,547,363   
Huang High School        District          2,917          $1,910,635   
Hernandez High School    District          4,635          $3,022,020   
Johnson High School      District          4,761          $3,094,650   

                       Per student budget Average math score  \
school_name                                                    
Figueroa High School                639.0               76.7   
Rodriguez High School               637.0               76.8   
Huang High School                   655.0               76.6   
Hernandez High School               652.0               77.3   
Johnson High School                 650.0               77.1   

                      Average reading score % passing math % passing reading  \
school_name                                                                    
Figueroa High School                   81.2          88.4%            100.0%   
Rodriguez High School                  80.7          88.5%            100.0%   
Huang High School                      81.2          88.9%            100.0%   
Hernandez High School                  80.9          89.1%            100.0%   
Johnson High School                    81.0          89.2%            100.0%   

                      % overall passing  
school_name                              
Figueroa High School              88.4%  
Rodriguez High School             88.5%  
Huang High School                 88.9%  
Hernandez High School             89.1%  
Johnson High School               89.2%

## Math and Reading Scores by Grade

they go in order of 10th, 11th, 12th, 9th.  

In [614]:
ninth_graders = data_complete[(data_complete["grade"] == "9th")]
tenth_graders = data_complete[(data_complete["grade"] == "10th")]
eleventh_graders = data_complete[(data_complete["grade"] == "11th")]
twelfth_graders = data_complete[(data_complete["grade"] == "12th")]

ninth_grade_math_scores = pd.DataFrame(ninth_graders.groupby(["school_name","grade"]).mean())
tenth_grade_math_scores = pd.DataFrame(tenth_graders.groupby(["school_name","grade"]).mean())
eleventh_grade_math_scores = pd.DataFrame(eleventh_graders.groupby(["school_name","grade"]).mean())
twelfth_grade_math_scores = pd.DataFrame(twelfth_graders.groupby(["school_name","grade"]).mean())

columns_to_drop = ["reading_score","Student ID","School ID", "size", "budget"]

ninth_grade_math_scores = ninth_grade_math_scores.drop(columns_to_drop, axis=1)
ninth_grade = ninth_grade_math_scores.rename(columns = {"math_score":"9th"})

tenth_grade_math_scores = tenth_grade_math_scores.drop(columns_to_drop, axis=1)
tenth_grade = tenth_grade_math_scores.rename(columns = {"math_score":"10th"})

eleventh_grade_math_scores = eleventh_grade_math_scores.drop(columns_to_drop, axis=1)
eleventh_grade = eleventh_grade_math_scores.rename(columns = {"math_score":"11th"})

twelfth_grade_math_scores = twelfth_grade_math_scores.drop(columns_to_drop, axis=1)
twelfth_grade = twelfth_grade_math_scores.rename(columns = {"math_score":"12th"})

ninth_grade = ninth_grade.droplevel("grade")
tenth_grade = tenth_grade.droplevel("grade")
eleventh_grade = eleventh_grade.droplevel("grade")
twelfth_grade = twelfth_grade.droplevel("grade")

math_merge1 = pd.merge(ninth_grade, tenth_grade, on="school_name")
math_merge2 = pd.merge(math_merge1,eleventh_grade, on="school_name")
math_scores_by_grade = pd.merge(math_merge2, twelfth_grade, on="school_name")
math_scores_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     77.083676  76.996772  77.515588  76.492218
Cabrera High School    83.094697  83.154506  82.765560  83.277487
Figueroa High School   76.403037  76.539974  76.884344  77.151369
Ford High School       77.361345  77.672316  76.918058  76.179963
Griffin High School    82.044010  84.229064  83.842105  83.356164
Hernandez High School  77.438495  77.337408  77.136029  77.186567
Holden High School     83.787402  83.429825  85.000000  82.855422
Huang High School      77.027251  75.908735  76.446602  77.225641
Johnson High School    77.187857  76.691117  77.491653  76.863248
Pena High School       83.625455  83.372000  84.328125  84.121547
Rodriguez High School  76.859966  76.612500  76.395626  77.690748
Shelton High School    83.420755  82.917411  83.383495  83.778976
Thomas High School     83.590022  83.087886  83.498795  83.497041
Wilson High School     83.085578  83.724422  83.195326  83.035794
Wright High School     83.264706  84.010288  83.836782  83.644986

In [615]:
ninth_grade_reading_scores = pd.DataFrame(ninth_graders.groupby(["school_name","grade"]).mean())
tenth_grade_reading_scores = pd.DataFrame(tenth_graders.groupby(["school_name","grade"]).mean())
eleventh_grade_reading_scores = pd.DataFrame(eleventh_graders.groupby(["school_name","grade"]).mean())
twelfth_grade_reading_scores = pd.DataFrame(twelfth_graders.groupby(["school_name","grade"]).mean())

columns_to_drop = ["math_score","Student ID","School ID", "size", "budget"]

ninth_grade_reading_scores = ninth_grade_reading_scores.drop(columns_to_drop, axis=1)
ninth_grader = ninth_grade_reading_scores.rename(columns = {"reading_score":"9th"})

tenth_grade_reading_scores = tenth_grade_reading_scores.drop(columns_to_drop, axis=1)
tenth_grader = tenth_grade_reading_scores.rename(columns = {"reading_score":"10th"})

eleventh_grade_reading_scores = eleventh_grade_reading_scores.drop(columns_to_drop, axis=1)
eleventh_grader = eleventh_grade_reading_scores.rename(columns = {"reading_score":"11th"})

twelfth_grade_reading_scores = twelfth_grade_reading_scores.drop(columns_to_drop, axis=1)
twelfth_grader = twelfth_grade_reading_scores.rename(columns = {"reading_score":"12th"})

ninth_grader = ninth_grader.droplevel("grade")
tenth_grader = tenth_grader.droplevel("grade")
eleventh_grader = eleventh_grader.droplevel("grade")
twelfth_grader = twelfth_grader.droplevel("grade")

reading_merge1 = pd.merge(ninth_grader, tenth_grader, on="school_name")
reading_merge2 = pd.merge(reading_merge1,eleventh_grader, on="school_name")
reading_scores_by_grade = pd.merge(reading_merge2, twelfth_grader, on="school_name")
reading_scores_by_grade

9th       10th       11th       12th
school_name                                                      
Bailey High School     81.303155  80.907183  80.945643  80.912451
Cabrera High School    83.676136  84.253219  83.788382  84.287958
Figueroa High School   81.198598  81.408912  80.640339  81.384863
Ford High School       80.632653  81.262712  80.403642  80.662338
Griffin High School    83.369193  83.706897  84.288089  84.013699
Hernandez High School  80.866860  80.660147  81.396140  80.857143
Holden High School     83.677165  83.324561  83.815534  84.698795
Huang High School      81.290284  81.512386  81.417476  80.305983
Johnson High School    81.260714  80.773431  80.616027  81.227564
Pena High School       83.807273  83.612000  84.335938  84.591160
Rodriguez High School  80.993127  80.629808  80.864811  80.376426
Shelton High School    84.122642  83.441964  84.373786  82.781671
Thomas High School     83.728850  84.254157  83.585542  83.831361
Wilson High School     83.939778  84.021452  83.764608  84.317673
Wright High School     83.833333  83.812757  84.156322  84.073171

## Scores by School Spending

In [ ]:
spending_bins = [0,585,630,645,680]
labels = ["<$585", "$585-$630", "$630-$645", "$645-$680"]

organized_df["Spending Ranges(Per Student)"]=pd.cut(organized_df["Per student budget"],spending_bins, labels=labels)

spending_math_scores = organized_df.groupby(["Spending Ranges(Per Student)"])["Average math score"].mean()
spending_reading_scores = organized_df.groupby(["Spending Ranges(Per Student)"])["Average reading score"].mean()
spending_passing_math = organized_df.groupby(["Spending Ranges(Per Student)"])["% passing math"].mean()
spending_passing_reading = organized_df.groupby(["Spending Ranges(Per Student)"])["% passing reading"].mean()
overall_passing_spending = organized_df.groupby(["Spending Ranges(Per Student)"])["% overall passing"].mean()





In [ ]:
spend_merge_1 = pd.merge(spending_math_scores, spending_reading_scores, on="Spending Ranges(Per Student)")
spend_merge_2 = pd.merge(spend_merge_1, spending_passing_math, on="Spending Ranges(Per Student)")
spend_merge_3 = pd.merge(spend_merge_2, spending_passing_reading, on="Spending Ranges(Per Student)")
spending_summary = pd.merge(spend_merge_3,overall_passing_spending, on = "Spending Ranges(Per Student)")


map(format) numerical values in this completed data frame

In [ ]:
spending_summary["% passing reading"] = spending_summary["% passing reading"].map("{:.1f}%".format)
spending_summary["% passing math"] = spending_summary["% passing math"].map("{:.1f}%".format)
spending_summary["% overall passing"] = spending_summary["% overall passing"].map("{:.1f}%".format)
spending_summary["Average reading score"] = spending_summary["Average reading score"].map("{:.1f}".format)
spending_summary["Average math score"] = spending_summary["Average math score"].map("{:.1f}".format)

spending_summary

## Scores by School Size

In [ ]:
size_bins = [0,1000,2000,5000]

labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

organized_df["Total Students Ranges"]=pd.cut(organized_df["Total Students"],size_bins, labels=labels)

size_math_scores = organized_df.groupby(["Total Students Ranges"])["Average math score"].mean()
size_reading_scores = organized_df.groupby(["Total Students Ranges"])["Average reading score"].mean()
size_passing_math = organized_df.groupby(["Total Students Ranges"])["% passing math"].mean()
size_passing_reading = organized_df.groupby(["Total Students Ranges"])["% passing reading"].mean()
size_passing_overall = organized_df.groupby(["Total Students Ranges"])["% overall passing"].mean()
size_student_budget = organized_df.groupby(["Total Students Ranges"])["Per student budget"].mean()

In [ ]:
size_merge_1 = pd.merge(size_math_scores, size_reading_scores, on="Total Students Ranges")
size_merge_2 = pd.merge(size_merge_1, size_passing_math, on="Total Students Ranges")
size_merge_3 = pd.merge(size_merge_2, size_passing_reading, on="Total Students Ranges")
size_merge_4 = pd.merge(size_merge_3,size_student_budget, on = "Total Students Ranges")
size_summary = pd.merge(size_merge_4,size_passing_overall, on = "Total Students Ranges")

In [ ]:
size_summary["% passing reading"] = size_summary["% passing reading"].map("{:.1f}%".format)
size_summary["% passing math"] = size_summary["% passing math"].map("{:.1f}%".format)
size_summary["% overall passing"] = size_summary["% overall passing"].map("{:.1f}%".format)
size_summary["Average reading score"] = size_summary["Average reading score"].map("{:.1f}".format)
size_summary["Average math score"] = size_summary["Average math score"].map("{:.1f}".format)
size_summary["Per student budget"] = size_summary["Per student budget"].map("${:.2f}".format)

## Scores by School Type

In [ ]:
type_math_scores = organized_df.groupby(["School Type"])["Average math score"].mean()
type_reading_scores = organized_df.groupby(["School Type"])["Average reading score"].mean()
type_passing_math = organized_df.groupby(["School Type"])["% passing math"].mean()
type_passing_reading = organized_df.groupby(["School Type"])["% passing reading"].mean()
type_passing_overall = organized_df.groupby(["School Type"])["% overall passing"].mean()
type_student_budget = organized_df.groupby(["School Type"])["Per student budget"].mean()

In [ ]:
type_merge_1 = pd.merge(type_math_scores, type_reading_scores, on="School Type")
type_merge_2 = pd.merge(type_merge_1, type_passing_math, on="School Type")
type_merge_3 = pd.merge(type_merge_2, type_passing_reading, on="School Type")
type_merge_4 = pd.merge(type_merge_3, type_student_budget, on= "School Type")
type_summary = pd.merge(type_merge_4,type_passing_overall, on = "School Type")

In [ ]:
type_summary["% passing reading"] = type_summary["% passing reading"].map("{:.1f}%".format)
type_summary["% passing math"] = type_summary["% passing math"].map("{:.1f}%".format)
type_summary["% overall passing"] = type_summary["% overall passing"].map("{:.1f}%".format)
type_summary["Average reading score"] = type_summary["Average reading score"].map("{:.1f}".format)
type_summary["Average math score"] = type_summary["Average math score"].map("{:.1f}".format)
type_summary["Per student budget"] = type_summary["Per student budget"].map("${:.2f}".format)

In [ ]:
organized_df["Total Students"] = organized_df["Total Students"].map("{:,}".format)
organized_df["Total school budget"] = organized_df["Total school budget"].map("${:,}".format)
organized_df["Per student budget"] = organized_df["Per student budget"].map("${:.0f}".format)
organized_df["Average math score"] = organized_df["Average math score"].map("{:.1f}".format)
organized_df["Average reading score"] = organized_df["Average reading score"].map("{:.1f}".format)
organized_df["% passing reading"] = organized_df["% passing reading"].map("{:.1f}%".format)